### ISSUE TITLE
False positive Argument 2 to "get" of "Message" has incompatible type "str"; expected "bool"

### ISSUE THREAD

**Issue by charlax** [python/mypy#7192]
Note: I wasn't sure what a proper title was for this issue, so I chose one that is specific to the issue at hand. I checked `typeshed`, it looks like the issue is about mypy, not typeshed. I wouldn't be surprised if an issue already existed for this false positive.

Minimal repro:

```python
assert isinstance(email, EmailMessage)

# Fine
r1 = email.get("List-Id", "")
# Fine
r2 = email.get("List-Id", "") or email.get("List-Unsubscribe", "")
# Fine
r3 = bool(email.get("List-Id", "") or email.get("List-Unsubscribe", ""))
# FAILS: Argument 2 to "get" of "Message" has incompatible type "str"; expected "bool"
r4 = bool(
    email.get("Auto-Submitted", "") in ("no", "")
    or email.get("List-Unsubscribe", "")
)
# FAILS: Argument 2 to "get" of "Message" has incompatible type "str"; expected "bool"
r5 = email.get("Auto-Submitted", "") in ("no", "") or email.get(
    "List-Unsubscribe", ""
)
```

Both are false positives.

Typeshed:

```python
_T = TypeVar('_T')

class Message:
    ...
    def get(self, name: str, failobj: _T = ...) -> Union[_HeaderType, _T]: ...
```

* What is the actual behavior/output? `Argument 2 to "get" of "Message" has incompatible type "str"; expected "bool"`
* What is the behavior/output you expect? No error.
* What are the versions of mypy and Python you are using? mypy 0.711, Python 3.6.7
  Do you see the same issue after installing mypy from Git master? haven't tried it but since 0.711 is fairly recent...

* What are the mypy flags you are using? (For example --strict-optional)

```
[mypy]
ignore_missing_imports = True
follow_imports = skip
check_untyped_defs = False
strict_equality = True
disallow_incomplete_defs = True
```

**Comment by msullivan** [comment]
Looks like the return context for the second call is getting inferred as `bool` which is constraining the result.

Awkward.

**Comment by collinanderson** [comment]
Yes, it seems mypy looks at where the output of .get() is going ("return context"), and expects the default argument to match that type, which is backwards from what I would expect.

So if it's being used within boolean logic, mypy expects the default value to be boolean.
`bool(something or email.get("List-Unsubscribe", ""))`
A workaround is to explicitly wrap `.get()` in `bool()`, because bool accepts `object` (anything).
`bool(something or bool(email.get("List-Unsubscribe", "")))`

The error message is really confusing if you're assigning to the wrong type: it complains about the default value, when the real issue is you're assigning to the wrong type:
```
a: int
a = email.get("List-Unsubscribe", "")
```
The above will complain about the default arg not being `int`, when mypy should instead be complaining about the output of `.get()` being `str` and the variable being `int`.
